grpc-gcp/src/main/java/com/google/cloud/grpc/multiendpoint/MultiEndpoint.java

```diff
@@ -0,0 +1,201 @@
+ * endpoint for up to recovery timeout after it became unavailable to give it some time to recover.
+ *
+ * <p>The list of endpoints can be changed at any time with {@link #setEndpoints} method.
+ * MultiEndpoint will preserve endpoints' state and update their priority according to their new
+ * positions.
+ *
+ * <p>The initial state of endpoint is "unavailable" or "recovering" if using recovery timeout.
+ */
+@CheckReturnValue
+public final class MultiEndpoint {
+  @GuardedBy("this")
+  private final Map<String, Endpoint> endpointsMap = new HashMap<>();
+
+  @GuardedBy("this")
+  private String currentId;
+
+  private final Duration recoveryTimeout;
+  private final boolean recoveryEnabled;
+
+  private final ScheduledThreadPoolExecutor executor = new ScheduledThreadPoolExecutor(1);
+
+  private MultiEndpoint(Builder builder) {
+    this.recoveryTimeout = builder.recoveryTimeout;
+    this.recoveryEnabled =
+        !builder.recoveryTimeout.isNegative() && !builder.recoveryTimeout.isZero();
+    this.setEndpoints(builder.endpoints);
+  }
+
+  /** Builder for MultiEndpoint. */
+  public static final class Builder {
+    private final List<String> endpoints;
+    private Duration recoveryTimeout = Duration.ZERO;
+
+    public Builder(List<String> endpoints) {
+      Preconditions.checkNotNull(endpoints);
+      Preconditions.checkArgument(!endpoints.isEmpty(), "Endpoints list must not be empty.");
+      this.endpoints = endpoints;
+    }
+
+    /**
+     * MultiEndpoint will keep the current endpoint for up to recovery timeout after it became
+     * unavailable to give it some time to recover.
+     */
+    public Builder withRecoveryTimeout(Duration timeout) {
+      Preconditions.checkNotNull(timeout);
+      this.recoveryTimeout = timeout;
+      return this;
+    }
+
+    public MultiEndpoint build() {
+      return new MultiEndpoint(this);
+    }
+  }
+
+  /**
+   * Returns current endpoint id.
+   *
+   * <p>Note that the read is not synchronized and in case of a race condition there is a chance of
+   * getting an outdated current id.
+   */
+  @SuppressWarnings("GuardedBy")
+  public String getCurrentId() {
+    return currentId;
+  }
+
+  private synchronized void setEndpointStateInternal(String endpointId, EndpointState state) {
+    Endpoint endpoint = endpointsMap.get(endpointId);
+    if (endpoint != null) {
+      endpoint.setState(state);
+      maybeFallback();
+    }
+  }
+
+  /** Inform MultiEndpoint when an endpoint becomes available or unavailable. */
+  public synchronized void setEndpointAvailable(String endpointId, boolean available) {
+    setEndpointState(endpointId, available ? EndpointState.AVAILABLE : EndpointState.UNAVAILABLE);
+  }
+
+  private synchronized void setEndpointState(String endpointId, EndpointState state) {
+    Preconditions.checkNotNull(state);
+    Endpoint endpoint = endpointsMap.get(endpointId);
+    if (endpoint == null) {
+      return;
```

Comment:
No, it's just because we can get a `setEndpointState` for an endpoint we don't have. E.g., the list of endpoint was updated and we removed few endpoints and then we received a state update for them. We are just not interested in state updates for endpoints we don't have.